daemon/cmd/devices.go

```diff
@@ -0,0 +1,257 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2016-2021 Authors of Cilium
+
+// This module implements Cilium's network device detection.
+
+package cmd
+
+import (
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/datapath/linux/probes"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/mac"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+)
+
+var (
+	excludedDevicePrefixes = []string{
+		"cilium_",
+		"lo",
+		"lxc",
+		"cni",
+		"docker",
+	}
+)
+
+type DeviceManager struct {
+	mu      lock.Mutex
+	devices map[string]bool
+	l3DevOK bool
+}
+
+func NewDeviceManager() *DeviceManager {
+	dm := &DeviceManager{
+		l3DevOK: supportL3Dev(),
+		devices: make(map[string]bool),
+	}
+	return dm
+}
+
+func (dm *DeviceManager) Close() {
+}
+
+// Detect tries to detect devices which are to be used for:
+// NodePort BPF, direct routing in NodePort BPF, IPsec and Bandwidth Manager.
+//
+// The devices are detected by looking at all the configured global unicast
+// routes in the system.
+func (dm *DeviceManager) Detect() error {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+
+	if err := expandDevices(); err != nil {
+		log.WithError(err).Warnf("Failed to expand device wildcards")
+	}
+
+	// Detect the devices from the system routing table by picking the devices
+	// which have global unicast routes.
+	if len(option.Config.Devices) == 0 {
+		routes, err := netlink.RouteList(nil, netlink.FAMILY_ALL)
+		if err != nil {
+			return fmt.Errorf("cannot retrieve routes for device detection: %w", err)
+		}
+		dm.updateDevicesFromRoutes(routes)
+	} else {
+		for _, dev := range option.Config.Devices {
+			dm.devices[dev] = true
+		}
+	}
+
+	if option.Config.DirectRoutingDevice != "" {
+		dm.devices[option.Config.DirectRoutingDevice] = true
+	}
+
+	if option.Config.IPv6MCastDevice != "" {
+		dm.devices[option.Config.IPv6MCastDevice] = true
+	}
+
+	detectDirectRoutingDev := option.Config.EnableNodePort && option.Config.DirectRoutingDevice == ""
+	detectIPv6MCastDev := option.Config.EnableIPv6NDP && len(option.Config.IPv6MCastDevice) == 0
+	if detectDirectRoutingDev || detectIPv6MCastDev {
+		nodeIP := node.GetK8sNodeIP()
+
+		if nodeIP == nil {
+			return fmt.Errorf("Cannot determine direct routing or IPv6 multicast device as K8s node IP is not known")
+		} else {
+			for dev := range dm.devices {
+				link, err := netlink.LinkByName(dev)
+				if err != nil {
+					return fmt.Errorf("Cannot find device '%s': %w", dev, err)
+				}
+
+				// Check if any of the addresses assigned to this device is the k8s node IP and if so
+				// use it for direct routing and IPv6 multicast.
+				if addrs, err := netlink.AddrList(link, netlink.FAMILY_ALL); err == nil {
+					for _, a := range addrs {
+						if a.IP.Equal(nodeIP) {
```

Comment:
Could we add the dev to the list of devs in any case if it matches k8s node ip (given here we know the intent for sure)?

Reply:
Done.